### PR TITLE
Opfibrations

### PR DESCRIPTION
**Pull request by TOTBWF**
# Description

This PR is a bit beefy, so it's best to think of it in chunks. First, it proves a bunch of duality properties of morphism classes (IE: monos in `C^op` are epis in `C`, etc). Next, it defines the total opposite category, and proves duality properties for *displayed* morphism classes. Finally, it defines cocartesian maps and opfibrations, and proves a bunch of properties about them, mostly using duality arguments.

### CODE REVIEW

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+*covariant* pseudofunctorial families of categories? Such pseudofunctors
+can be encoded by dualising cartesian fibrations.
+
+[Cartesian fibrations]: Cat.Displayed.Cartesian.html
+
+To do this, we must first dualise the notion of a cartesian map to a
+**cocartesian map**. Fix a map $a \to b$ in $\ca{B}$, objects $a'$
+and $b'$ displayed over $a$ and $b$ resp., and a map $f' : a' \to_f b'$
+over $f$. We say that $f'$ is **cocartesian** if it has the shape of a
+"pushout diagram", in contrast to the cartesian maps "pullback diagrams".
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:36]: Should this be "cartesian maps, which are ..."? Or "cartesian maps' ..."?

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    universal′ (is-set→squarep (λ _ _ → Hom-set a u) (ap (_∘ f) q) p r refl i) h′
+
+  uniquep : ∀ {u u′} {m₁ m₂ : Hom b u} {k : Hom a u}
+          → (p : m₁ ∘ f ≡ k) (q : m₁ ≡ m₂) (r : m₂ ∘ f ≡ k)
+          → {h′ : Hom[ k ] a′ u′}
+          → (m′ : Hom[ m₁ ] b′ u′)
+          → m′ ∘′ f′ ≡[ p ] h′ → m′ ≡[ q ] universal′ r h′
+  uniquep p q r {h′ = h′} m′ s  =
+    to-pathp⁻ (unique m′ (from-pathp⁻ s) ∙ from-pathp⁻ (universalp p q r h′))
+```
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:125]: I feel like this whole section could be omitted from the website instead...

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+## Duality
+
+As noted before, cocartesian maps are dual to cartesian maps. We
+can make this correspondence precise by showing that cartesian maps
+in the total opposite of $\ca{E}$ are cocartesian maps, and vice versa.
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:131]: Link to [total opposite]?

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+```
+
+## Properties of Cocartesian Morphisms
+
+<details>
+<summary>
+We shall now prove some properties of cocartesian morphisms. These
+are all dual to the properties that we've proved about cartesian
+morphisms, so we will appeal to duality to prove them!
+</summary>
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:195]: Can you like.. pull the type signatures out of the `<details>` block, having this explanation in a stand-alone paragraph?

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+    lifting : Hom[ f ] x′ y′
+    cocartesian : Cocartesian f lifting
+  open Cocartesian cocartesian public
+```
+
+<details>
+<summary>
+We also can apply duality to cocartesian lifts.
+</summary>
+```agda
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:298]: Needs a line break here

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+  Cocartesian-lift.y′ cocart
+cocartesian-lift→cartesian-lift^op cocart .Cartesian-lift.lifting =
+  Cocartesian-lift.lifting cocart
+cocartesian-lift→cartesian-lift^op cocart .Cartesian-lift.cartesian =
+  cocartesian→cartesian^op (Cocartesian-lift.cocartesian cocart)
+```
+</details>
+
+We can use this notion to define cocartesian fibrations (sometimes
+referred to as opfibrations).
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:324]:
```suggestion
We can use this notion to define cocartesian fibrations (sometimes
referred to as **opfibrations**).
```

File: src/Cat/Displayed/Cocartesian/Weak.lagda.md [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+  → Weak-cocartesian f f′
+  → Weak-cartesian (ℰ ^total-op) f f′
+weak-cocartesian→weak-cartesian^op wcocart .Weak-cartesian.universal =
+  Weak-cocartesian.universal wcocart
+weak-cocartesian→weak-cartesian^op wcocart .Weak-cartesian.commutes =
+  Weak-cocartesian.commutes wcocart
+weak-cocartesian→weak-cartesian^op wcocart .Weak-cartesian.unique =
+  Weak-cocartesian.unique wcocart
+```
+<details>
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian/Weak.lagda.md:84]: This is an opening tag!

File: src/Cat/Displayed/Cocartesian/Weak.lagda.md [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+  → Cocartesian-fibration
+  → Weak-cocartesian f f′
+  → Cocartesian f f′
+weak-cocartesian→cocartesian opfib wcocart =
+  cartesian^op→cocartesian $
+  weak-cartesian→cartesian (ℰ ^total-op)
+    (opfibration→fibration^op opfib)
+    (weak-cocartesian→weak-cartesian^op wcocart)
+```
+<details>
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian/Weak.lagda.md:125]: This is also an opening tag!

File: src/Cat/Displayed/Cocartesian/Weak.lagda.md [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+weak-cocartesian→weak-cartesian^op wcocart .Weak-cartesian.universal =
+  Weak-cocartesian.universal wcocart
+weak-cocartesian→weak-cartesian^op wcocart .Weak-cartesian.commutes =
+  Weak-cocartesian.commutes wcocart
+weak-cocartesian→weak-cartesian^op wcocart .Weak-cartesian.unique =
+  Weak-cocartesian.unique wcocart
+```
+<details>
+
+## Properties
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian/Weak.lagda.md:86]: I would remove the section break between these.. and pull out the type signatures to outside the `<details>` block

File: src/Cat/Displayed/Cocartesian/Weak.lagda.md [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+	&&&& u \\
+	x && y
+	\arrow[lies over, from=2-1, to=4-1]
+	\arrow[lies over, from=2-3, to=4-3]
+	\arrow["{f'}"{description}, from=2-1, to=2-3]
+	\arrow["f"{description}, from=4-1, to=4-3]
+	\arrow["m", from=4-3, to=3-5]
+	\arrow[color={rgb,255:red,92;green,214;blue,92}, dashed, from=2-3, to=1-5]
+	\arrow["{h'}", curve={height=-30pt}, from=2-1, to=1-5]
+	\arrow[lies over, from=1-5, to=3-5]
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian/Weak.lagda.md:158]: Could this be two spaces?

File: src/Cat/Displayed/Cocartesian/Weak.lagda.md [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+    {y′}    : Ob[ y ]
+    lifting : Hom[ f ] x′ y′
+    weak-cocartesian : Weak-cocartesian f lifting
+
+  open Weak-cocartesian weak-cocartesian public
+```
+
+<details>
+<summary> As expected, weak cocartesian lifts are dual to weak cartesian lifts.
+</summary>
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian/Weak.lagda.md:390]: Add a blank line here, ditto for lifting the type signatures

File: src/Cat/Displayed/Cocartesian/Weak.lagda.md [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+	\arrow["{h^{**}}", curve={height=-6pt}, from=2-3, to=1-3]
+	\arrow["{id^{*}}"', color={rgb,255:red,214;green,92;blue,92}, curve={height=6pt}, from=2-3, to=1-3]
+\end{tikzcd}
+~~~
+
+However, $h^{**}$ is the *unique* vertical map that factorises $f'$
+through $h^{*}$, so it suffices to show that the cell highlighted in
+blue commutes.
+
+~~~{.quiver}
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian/Weak.lagda.md:324]: I think all the diagrams in this page could benefit from being

```
~~~{.quiver .tall-2}
```

+ replace the hard tabs for two spaces

File: src/Cat/Displayed/Instances/Family.lagda.md [hunk 13]
```diff
@@ -89,6 +89,53 @@ Family-is-cartesian = iscart where
   iscart .has-lift {x = x} {y} f y′ .cartesian = cart {x = x} {y} f y′
 ```
 
+Morphisms in the family fibration are cartesian if and only if they are
+point-wise isomorphisms. Showing the forward direction is a matter of
```
Comment (plt-amy) [inline src/Cat/Displayed/Instances/Family.lagda.md:93]:
```suggestion
pointwise isomorphisms. Showing the forward direction is a matter of
```

File: src/Cat/Displayed/Morphism.lagda.md [hunk 14]
```diff
@@ -95,14 +95,16 @@ is-weak-monic-is-prop f′ mono mono′ i g′ g″ p =
   is-prop→pathp (λ i → Hom[ _ ]-set _ _ g′ g″)
     (mono g′ g″ p) (mono′ g′ g″ p) i
 
-record _↪ʷ_
+record _↪[_]ʷ_
```
Comment (plt-amy) [inline src/Cat/Displayed/Morphism.lagda.md:98]: I'm starting to hate these names. Thoughts on `weak-mono-over`?

File: src/Cat/Displayed/Morphism.lagda.md [hunk 15]
```diff
@@ -371,6 +377,26 @@ make-iso[ inv ] f′ g′ p q .from′ = g′
 make-iso[ inv ] f′ g′ p q .inverses′ .Inverses[_].invl′ = p
 make-iso[ inv ] f′ g′ p q .inverses′ .Inverses[_].invr′ = q
 
+make-vert-iso
```
Comment (plt-amy) [inline src/Cat/Displayed/Morphism.lagda.md:380]:
```suggestion
make-vertical-iso
```

File: src/Cat/Displayed/Morphism/Duality.lagda.md [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+case, just mountains of boilerplate.
+
+We start by showing that displayed monos and epis are dual.
+
+```agda
+monic[]^op→epic[]
+  : ∀ {f : a ℬ^op.↪ b}
+  → a′ ℰ^op.↪[ f ] b′
+  → b′ ℰ.↠[ monic^op→epic f ] a′
+monic[]^op→epic[] f .Cat.Displayed.Morphism.mor′ =
```
Comment (plt-amy) [inline src/Cat/Displayed/Morphism/Duality.lagda.md:53]: Definitely open all of these records so you don't have to repeat the module name in the copattern

File: src/Cat/Displayed/Reasoning.lagda.md [hunk 17]
```diff
@@ -58,6 +58,17 @@ reindex
 reindex p q {f′} = ap (λ e → hom[ e ] f′) (B.Hom-set _ _ _ _ p q)
 ```
 
+```agda
+cast[]
```
Comment (plt-amy) [inline src/Cat/Displayed/Reasoning.lagda.md:62]:
```suggestion

cast[]
```

Comment (plt-amy) [reply src/Cat/Displayed/Reasoning.lagda.md:62]: (if the suggestion is wonky, I mean "put them in the same code block")

File: src/Cat/Displayed/Total/Op.lagda.md [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+open Functor
+open Total-hom
+```
+
+# Total Opposites
+
+Opposites of displayed categories are somewhat subtle, as there are
+multiple constructions that one could reasonably call the "opposite".
+The most obvious construction is to construct a new
+displayed category over $\ca{B}^{op}$; we call this category the
```
Comment (plt-amy) [inline src/Cat/Displayed/Total/Op.lagda.md:21]:
```suggestion
displayed category over $\ca{B}\op$; we call this category the
```

File: src/Cat/Displayed/Morphism/Duality.lagda.md [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+```
+-->
+
+
+# Duality of Displayed Morphism Classes
+
+In this file we prove that morphism classes in a displayed category
+correspond to their duals in the total opposite displayed category.
+There is *even less* mathematical content here than the non-displayed
+case, just mountains of boilerplate.
```
Comment (plt-amy) [inline src/Cat/Displayed/Morphism/Duality.lagda.md:44]: Make "non-displayed case" a link?

File: src/Cat/Displayed/Cartesian.lagda.md [hunk 20]
```diff
@@ -25,7 +25,7 @@ _collection_ of fibre categories $\ca{E}^*(x)$, this assignment isn't
 necessarily functorial!
 
 More precisely, we should have that a collection of categories, indexed
-by a category, should be a _pseudofunctor_ $\ca{B} \to \Cat$, where
+by a category, should be a _pseudofunctor_ $\ca{B}^{op} \to \Cat$, where
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian.lagda.md:28]:
```suggestion
by a category, should be a _pseudofunctor_ $\ca{B}\op \to \Cat$, where
```

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+  \\
+  a && b
+  \arrow["{f'}"', from=1-1, to=1-3]
+  \arrow["f", from=3-1, to=3-3]
+  \arrow[lies over, from=1-1, to=3-1]
+  \arrow[lies over, from=1-3, to=3-3]
+\end{tikzcd}\]
+~~~
+
+```agda
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:50]: So to fit this should be called `is-cocartesian`, same for `is-cartesian` in the other module. Thoughts?

File: src/Cat/Displayed/Cocartesian.lagda.md [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+```
+
+## Duality
+
+As noted before, cocartesian maps are dual to cartesian maps. We
+can make this correspondence precise by showing that cartesian maps
+in the total opposite of $\ca{E}$ are cocartesian maps, and vice versa.
+
+```agda
+cartesian^op→cocartesian
```
Comment (plt-amy) [inline src/Cat/Displayed/Cocartesian.lagda.md:134]: To fit with Cat.Diagram.Duals this should be called `co-cartesian→cocartesian`

**Review by plt-amy — CHANGES_REQUESTED** [review]

File: src/Cat/Displayed/Cocartesian.lagda.md (hunk 21 above)
Comment (TOTBWF) [reply src/Cat/Displayed/Cocartesian.lagda.md:50]: Agreed. It also removes a conflict from `Cat.Diagram.Product`, so easy win1

**Review by plt-amy — APPROVED** [review]

**Pull request merged**